feat: add serverless application example

vjcspy opened the pull request:
this pr add an example for serverless application. Some points of interest:

- work offline with `serverless-offline`
- bundle with `webpack`
- implement `jest` 
- support `dotenv`

NiGhTTraX commented:
Hey @vjcspy, thanks for your PR! Can you please import something from the monorepo in your app, like [foo/meaningOfLife](https://github.com/NiGhTTraX/ts-monorepo/blob/1eb83a2c6b6f6f9154d9255bed17e33831380ec2/packages/foo/src/index.ts#L1) and make sure it correctly uses the [path aliases](https://github.com/NiGhTTraX/ts-monorepo/blob/1eb83a2c6b6f6f9154d9255bed17e33831380ec2/tsconfig.json#L8)?

Also, you can remove the jest and dotenv stuff since I like to keep examples focused only on one thing. jest is already covered in [apps/jest](https://github.com/NiGhTTraX/ts-monorepo/tree/1eb83a2c6b6f6f9154d9255bed17e33831380ec2/apps/jest-tsjest).

If you make sure the monorepo imports work, I can take care of cleaning up the rest. Thanks!

vjcspy commented:
Thanks for you response. I refactored to make it as simple as possible. Removed jest + dotenv, also used alias path in handler.ts to import app.ts.
it works properly in local with command `pnpm -F serverless dev`

NiGhTTraX commented on apps/serverless/src/handler.ts:
```diff
@@ -0,0 +1,10 @@
+import type { Handler } from "serverless-http";
+import serverlessHttp from "serverless-http";
+
+import { app } from "@/app";
```
Importing "locally" is not enough to validate that monorepo aliases work. Please import `@nighttrax/foo` and return `meaningOfLife` in `requestListener` instead.

vjcspy replied:
fixed

NiGhTTraX commented on apps/serverless/src/app.ts:
```diff
@@ -0,0 +1,11 @@
+import { IncomingMessage, ServerResponse } from "http";
```
The `http` dependency is missing from `package.json`.

vjcspy replied:
fixed

NiGhTTraX commented on apps/serverless/package.json:
```diff
@@ -0,0 +1,29 @@
+    "serverless-http": "^3.0.2"
+  },
+  "devDependencies": {
+    "@types/supertest": "^2.0.12",
```
There are some unused deps here, left from when the example included jest and dotenv.

vjcspy replied:
fixed

NiGhTTraX reviewed (commented):
For some reason GitHub is not triggering the workflow for this PR, otherwise it would have caught a bunch of lint errors. Please run [all of these](https://github.com/NiGhTTraX/ts-monorepo/blob/698526b748348c5859618d62a7e3a80fe2635d7b/.github/workflows/tests.yml#L29-L36) locally and make sure they pass.

NiGhTTraX commented:
Thanks for updating this @vjcspy. I rebased the PR to solve lockfile conflicts and cleaned up a few things (you're going to have to do a force pull). There are still some issues however.

First of all, it looks like `serverless-bundle` has peer dependencies on React, which seems weird. I don't have any experience with the framework, but I don't see why React would be needed for lambda development.

```
apps/serverless
└─┬ serverless-bundle
  └─┬ isomorphic-style-loader
    ├── ✕ missing peer react@"^16.8.0 || ^17.0.0"
    └── ✕ missing peer react-dom@"^16.8.0 || ^17.0.0"
```

Secondly, running the `dev` script spews quite a lot of warnings when webpack tries to bundle `express`:

```
> cross-env NODE_ENV=development sls offline start --stage offline

Bundling with Webpack...
ERROR in ../../../../../../apps/serverless/node_modules/express/lib/application.js 16:19-42
Module not found: Error: Can't resolve 'finalhandler' in '/Users/nightcrawler/github/ts-monorepo/apps/serverless/node_modules/express/lib'
```

Trying to access the server at `localhost:4000` throws another error:

```
✖ Unhandled exception in handler 'app'.
✖ Error: Cannot find module 'body-parser'
✖ Runtime.ImportModuleError: Error: Cannot find module 'body-parser'
```

Can you please look into the above? Moreover, I would need some guidance on how to test this locally, and deploy it (to AWS I assume?). If I'm going to merge this example and maintain it, I need to understand what it does and how to work with it.

vjcspy commented:
Hi @NiGhTTraX, All of the above issues are caused by pnpm not installing express dependencies. You can check out express's package.json file and see that it relies a lot on other libraries (react, nextjs doesn't). There are 2 options to solve this problem:
1. Add settings to pnpm install peer dependencies. I prefer this option and find it necessary. (Nextjs is similar https://github.com/vercel/next.js/blob/canary/.npmrc)
2. If you still want to keep the settings of pnpm. Instead of using express I can switch back to using http module. It will not require external libraries.

Deploying to AWS is very simple. Follow this document to configure AWS profile, then just run 
```
pnpm -F @nighttrax/serverless deploy
```

NiGhTTraX commented:
@vjcspy [express](https://github.com/expressjs/express/blob/8368dc178af16b91b576c4c1d135f701a0007e5d/package.json) doesn't have any **peer** dependencies. Installing `express` with pnpm and webpack-ing it works out of the box, so there must be something off with `serverless`.

Also, what about `serverless-bundle`'s dependency on React?

vjcspy commented:
`serverless-bundle` config webpack resolve.symlinks = false. This is the cause of the error.
You can check it here `apps/serverless/node_modules/serverless-bundle/src/webpack.config.js`
Currently this package does not allow override webpack config. But try adding this setting for pnpm i see the issue is fixed
```
node-linker=hoisted
```
Also, `serverless-bundle` was used in [isomorphic-style-loader](https://github.com/kriasoft/isomorphic-style-loader) for supporting css and scss which depend on react.

I can switch to serverless-webpack. Actually, serverless-bundle uses the serverless-webpack plugin. Then I thought it would be better because we won't include unnecessary things.

NiGhTTraX commented:
@vjcspy if you can resolve the express issues and remove the React dependency then I'll take another look at this. It MUST work with pnpm's default config.
